### ISSUE TITLE
[Bug]  Cannot properly create database following the examples

### ISSUE THREAD

**Issue by roccazzella** [mariadb-operator/mariadb-operator#133]
**Describe the bug**
I'm following the example code in the Quickstart to instance a mariadb service + database.
I've successfully installed the operator in my cluster (via helm chart), and I can see all the CRDs.

I've created a new `mariadb` instance, I see it running and can connect (from within the `mariadb` pod and from a mysql client test pod in the same namespace).

I've then tried to created the a `database`, `user` and `grant` issue, but after creating them the status is 
`Ready: False, Status: Error connecting to MariaDB`

I cannot find a way to debug the problem (`kubectl describe` on the three objects doesn't give me any further detail).
I see there's also a `connection` CRD, but I cannot understand if that's something needed for the `database` instance to work (and how to configure it, secret, etc).

I could not find documentation on the `database`, `user` and `grant` CRDs, and cannot fully understand if they should create/maintain the respective SQL objects (eg, create a new database, create a user, etc) or they are just supposed to map entities that have to be independently created (database, user, grant, etc).

**Expected behaviour**
As showed in the main `README.md`, I expect to see three CR (`database`, `user`, `grant`) as `Ready: True, Status: Created`

**Steps to reproduce the bug**
  - install the operator thru helm chart
  - create a `mariadb` instance (image tag: 10.11.2)
  - verify that the service is up, and accepts connections (root user, from any pod whithin the namespace)
  - create a `database` instance
  
  **Further details**
  I didn't install init scripts (I don't need test tables/test data) nor backups.

**Environment details**:
- Kubernetes version: client: 1.22.3, server: 1.21.5
- mariadb-operator version: v0.0.15
- Install method: helm
- Install flavour: custom (recommended + my.cnf file in a configmap)

Thanks for your help.

**Comment by mmontes11** [comment]
Hey there @roccazzella ! thanks for raising this.

> I've then tried to created the a database, user and grant issue, but after creating them the status is
Ready: False, Status: Error connecting to MariaDB

These objects are created by the operator using the credentials specified in [spec.rootPasswordSecretKeyRef](https://github.com/mariadb-operator/mariadb-operator/blob/main/examples/manifests/mariadb_v1alpha1_mariadb.yaml#L6). Is this related to this PR maybe?:
- https://github.com/mariadb-operator/mariadb-operator/pull/134

If the readiness probes are failing, the `StatefulSet` won't be receiving traffic and therefore the operator won't be able to connect to MariaDB.

> I cannot find a way to debug the problem (kubectl describe on the three objects doesn't give me any further detail).
I see there's also a connection CRD, but I cannot understand if that's something needed for the database instance to work (and how to configure it, secret, etc).

The `Connection` CRD allows you to configure a connection for your application and provides the connection string as a `Secret`. 

Then, regarding debugging, we will be emitting `Events` that will help troubleshooting issues. They would appear in the `kubectl describe` output, see tracking issue:
- https://github.com/mariadb-operator/mariadb-operator/issues/8

 > I could not find documentation on the database, user and grant CRDs,
 
We are well aware that our documentation is minimal yet practical, that's why we are planning to launch a documentation site with a CRD API reference:
 - https://github.com/mariadb-operator/mariadb-operator/issues/21
 
 >  and cannot fully understand if they should create/maintain the respective SQL objects (eg, create a new database, create a user, etc) or they are just supposed to map entities that have to be independently created (database, user, grant, etc).
 
As I previously mentioned, the operator will take care of creating and maintaining these objects for you, not actually mapping them to existing ones in the MariaDB state. The aim is that you don't imperatively interact with MariaDB, you instead manage everything through CRDs in a GitOps fashion!
 
Does all of this make sense? 🙏🏻 
 
Thanks for your detailed feedback

**Comment by roccazzella** [comment]
hi @mmontes11!

and thanks for your detailed answer.

I have some more data (and context), I'll split my answer in two, hope I'm not too verbose : )

## TL;DR
We missed setting the correct `clusterName` in helm values when installing the operator, so any DNS-involving operation would fail.

### context for our `database`, `user`, `grant` (and probably also `connection`) CRs use-case

I failed to provide some better context for our use-case.

In our IDP we are mapping clients/projects to namespaces, and (currently, via simple Helm chart) instancing a single (project-wise global)  mariadb server for each namespace; each environment we create inside that namespace the (eg, live, stage, MR environments, etc) gets a specific database. 

Back to the operator: when instancing the general `mariadb.mmontes.io/v1alpha1./MariaDB` CR - and I set the `spec.database`, `spec.username`, ... values -  I see correctly that a database, user and grants are created (I can connect as `root` from within namespace pods), but I don't see the correspondent CRs. 

Our use-case would then be to instance environment-specific `database`, `user`, `grant` and `connection` CRs, to be able to efficiently manage environment-specific databases by creating all the relevant SQL items without directly instancing sql commands (eg, via the `root` user).
To give a practical example:
  - new project's CI/CD job can instance a `mariadb` CR > we get the project-wide db service
  - each environment's CI/CD can instance `database`, `user`, `grant` and `connection` CRs > the environment gets a database (and nice, neat k8s objects for us to use, great! :)

### back to my problem

When instancing environment-specific CRs (eg, for a `stage` database) we got a `Error connecting to MariaDB` status for all of them, despite the service (and root user access) being up and running (and in a working state).

I was able to dive a bit deeper into the operator's source code (despite my Go skills not being that good : ): I was able to spot the fault in our process that caused the error.

When installing the operator (via Helm chart) we followed the "Recommended installation" (hence activating metrics and certManager): that lead to the operator Deployment to be created with mostly default values, and specifically the `CLUSTER_NAME` env var being set to the default `clusterName` value (`cluster.localvalue`).

Unfortunately, our cluster has a specific name : )

Hence, `func ServiceFQDN` (inside `pkg/statefulset/statefulset.go`) would compose an invalid host, and from there the `Reconcile` function in `pkg/controller/sql/controller.go` would fail.

 By installing the operator with the correct CLUSTER_NAME value everything works as expected.
 
This was obviously our fault (not inspecting the default helm `values.yaml`, and fixing the correct cluster name in it), but it would maybe be useful to point that out in the Quickstart and/or in the new docs, to help future "distracted people" : )

Thanks again for your feedback, support and work on the operator!

**Comment by mmontes11** [comment]
Wow, thanks for sharing this @roccazzella. It's very valuable and exciting for us to see and understand how you are using `mariadb-operator` inside your organization. 🦭 ❤️ 

> We missed setting the correct clusterName in helm values when installing the operator, so any DNS-involving operation would fail.

First, congratulations on your in-depth investigation, this wasn't something obvious. As a takaway for us I would say that we need to document how to install `mariadb-operator` in a cluster with a non default domain, it's supported but not explicitly documented, thanks for digging this deep. This is a good topic for our [documentation site](https://github.com/mariadb-operator/mariadb-operator/issues/21)

> In our IDP we are mapping clients/projects to namespaces, and (currently, via simple Helm chart) instancing a single (project-wise global) mariadb server for each namespace;

I'm glad to see that a `MariaDB` per-namespace approach works in your org, it totally makes sense and it is the only approach supported. We were thinking about having a cluster wide `MariaDB` consumed by all clients/projects from all namespaces but this will require changes in our CRD design and opens the door to a lot of security concerns, that's why we decided to stick with the per-namespace approach for now.

> Back to the operator: when instancing the general mariadb.mmontes.io/v1alpha1./MariaDB CR - and I set the spec.database, spec.username, ... values - I see correctly that a database, user and grants are created (I can connect as root from within namespace pods), but I don't see the correspondent CRs.

This is the expected behaviour, if you specify `spec.database` and/or `spec.username` fields in the [MariaDB](https://github.com/mariadb-operator/mariadb-operator/blob/main/examples/manifests/mariadb_v1alpha1_mariadb.yaml#L10-L11) CRDs we will be leveraging `MARIADB_*` environment variables to create the initial user and database, see the [statefulset builder](https://github.com/mariadb-operator/mariadb-operator/blob/main/pkg/builder/statefulset_builder.go#L239-L258). There are some technical limitations behind this, but as a mitigation you can omit these fields in the `MariaDB` and create the initial [User](https://github.com/mariadb-operator/mariadb-operator/blob/main/examples/manifests/mariadb_v1alpha1_user.yaml) and [Database](https://github.com/mariadb-operator/mariadb-operator/blob/main/examples/manifests/mariadb_v1alpha1_database.yaml) as CRDs.

> new project's CI/CD job can instance a mariadb CR > we get the project-wide db service

Have you considered using GitOps for this? the CI/CD could potentially be creating a `MariaDB` CRD in a repo which will be eventually reconciled by your GitOps tool, see and example with [Flux](https://github.com/mariadb-operator/mariadb-operator/tree/main/examples/flux).

> and nice, neat k8s objects for us to use, great! :)

yes, exactly 🎉 

Thanks again for your detailed feedback 🙇🏻

**Comment by roccazzella** [comment]
hi @mmontes11 

and thanks for the further clarifications.

> This is the expected behaviour, if you specify spec.database and/or spec.username fields in the [MariaDB](https://github.com/mariadb-operator/mariadb-operator/blob/main/examples/manifests/mariadb_v1alpha1_mariadb.yaml#L10-L11) CRDs we will be leveraging MARIADB_* environment variables to create the initial user and database, see the [statefulset builder](https://github.com/mariadb-operator/mariadb-operator/blob/main/pkg/builder/statefulset_builder.go#L239-L258). There are some technical limitations behind this, but as a mitigation you can omit these fields in the MariaDB and create the initial [User](https://github.com/mariadb-operator/mariadb-operator/blob/main/examples/manifests/mariadb_v1alpha1_user.yaml) and [Database](https://github.com/mariadb-operator/mariadb-operator/blob/main/examples/manifests/mariadb_v1alpha1_database.yaml) as CRDs.

totally understandable, and the mitigation works for us: we'll just omit the `spec.database` values at service creation, and create the relevant objects at environment-level.


Thanks again for your support: looking forward to implement the operator in our CI/CD, and maybe - once my Go reaches a decent level - to try and implement some extra features on top of it (thinking of `database` creation as a clone of an existing one, data sync between extisting `database` objects... so many possibilities :)

I'll close the issue. Have a good day!
